evennia/contrib/slots_handler/slots.py

```diff
@@ -0,0 +1,385 @@
+    check `.db.slots`, `.ndb.slots`, and `.slots` in that order), and
+    `self.slots.attach()` will attempt to attach it on all of those slots (it
+    should fail if it can't attach on every slot). `self.slots.drop()` will
+    remove the target from any of the slots given it. Both of these methods
+    have the option to be given a custom `slots` argument, which will override
+    the slots on the held object.
+
+    Using the custom `slots` in `attach()` and `drop()` provides some
+    customization facility in that you can store *any* object in a slot, not
+    just an object that has been set up for it.
+    """
+
+    def __init__(self, obj):
+        self.obj = obj
+        self._objid = obj.id
+
+    # Internal utility methods
+    def is_name_valid(self, test):
+        "Internal function to check if the name is right."
+        try:
+            exec(test + " = 1")
+        except SyntaxError:
+            raise ValueError("You have to limit slot category names to "
+                             "characters that are valid in variable names.")
+
+    def defrag_nums(self, cats):
+        "Worker function to consolidate filled numbered slots."
+        if not isinstance(cats, list):
+            cats = [cats]
+
+        arrays = self.all(obj=True)
+        arrays = [a for a in arrays if a.key in cats]
+
+        for a in arrays:
+            slots = a.value
+            out = {k: v for k, v in slots.items() if isinstance(k, str)}
+            numbered = [(k, v) for k, v in slots.items() if isinstance(k, int)]
+            keys = [n[0] for n in numbered]
+            values = [n[1] for n in numbered]
+            self.obj.msg("Values: {}".format(values))
+            d = len(values) - 1
+            empty = [values.pop(0) for i in range(0, d) if values[0] == ""]
+            values = values + empty
+            numbered = zip(keys, values)
+            out.update(numbered)
+            self.obj.attributes.add(a.key, out, category="slots")
+
+    # Public methods
+    def all(self, obj=False):
+        "Return a dict of all slots."
+        d = self.obj.attributes.get(category="slots", return_obj=True)
+
+        if not d:
+            return {}
+        elif not isinstance(d, list):
+            d = [d]
+
+        if obj:
+            # Return attribute objects if requested.
+            return d
+        else:
+            # Return a dict detached from the database.
+            r = {s.key: s.value for s in d}
+            return r
+
+    def add(self, slots):
+        """
+        Create arrays of slots, or add additional slots to existing arrays.
+
+        Args:
+            slots: A dict of slots to add. Since you can't add empty
+                categories, it would be pointless to pass a list to this
+                function, and so it doesn't accept lists for input.
+        """
+
+        if not isinstance(slots, (dict, _SaverDict)):
+            return StatMsg(False, "You have to declare slots in the form "
```

Comment:
What is `StatMsg` and where is it defined/imported?

Reply:
Apologies. I've been playing around with a status message object and forgot to remove the references to it. I've changed all of those to `raise Exception()`.